lib/babe/verify.go

```diff
@@ -109,6 +112,54 @@ func (b *Session) verifyAuthorshipRight(slot uint64, header *types.Header) (bool
 		return false, fmt.Errorf("could not verify signature")
 	}
 
-	// TODO: check if the producer has equivocated, ie. have they produced a conflicting block?
+	// #553 check if the producer has equivocated, ie. have they produced a conflicting block?
+	hashes, err := b.blockState.GetAllHashesForParentDepth(header)
+	if err != nil {
+		return false, err
+	}
+
+	log.Debug("Got AllHashesForParentDepth", "len(hashes)", len(hashes))
+
+	for key, v := range hashes {
+		//get header
+		currentHeader, err := b.blockState.GetHeader(key)
+		if err != nil {
+			return false, fmt.Errorf("could not verify digests, key %s, depth, %d", key, v)
+		}
+
+		//compare digest
+		currentBlockProducerIndex := getBlockProducerIndex(currentHeader)
+		existingBlockProducerIndex := babeHeader.BlockProducerIndex
+
+		log.Debug("compare Digest", "currentBlockProducerIndex", currentBlockProducerIndex,
+			"existingBlockProducerIndex", existingBlockProducerIndex)
+
+		if reflect.DeepEqual(currentBlockProducerIndex, existingBlockProducerIndex) {
+			return false, fmt.Errorf("duplicated SealDigest")
+		}
+	}
+
 	return true, nil
 }
+
+func getBlockProducerIndex(header *types.Header) (blockProducerIndex uint64) {
+	preDigestBytes := header.Digest[0]
+
+	digestItem, err := types.DecodeDigestItem(preDigestBytes)
+	if err != nil {
+		return
```

Comment:
return error in this function, otherwise if there is an error it will return 0 which looks correct

Reply:
ok